src/core/presence.ts

```diff
@@ -350,31 +465,242 @@ export class DefaultPresence
-        clientId: member.clientId,
-        timestamp: member.timestamp,
-        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
-        data: member.data?.userCustomData as PresenceData,
+   * Converts a presence message to a presence member, with the action set to 'present'.
+   * @param presenceMessage
+   * @private
+   */
+  private _presenceMessageToPresenceMember(presenceMessage: Ably.PresenceMessage): InternalPresenceMember {
+    this._logger.trace('Presence._presenceMessageToPresenceMember();', { presenceMessage });
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+    const { clientId, timestamp, extras, data, connectionId } = presenceMessage;
+    const chatPresenceData = data as ChatPresenceData;
+    return {
+      connectionId,
+      clientId,
+      action: 'present' as PresenceEvents,
+      data: chatPresenceData.userCustomData,
+      updatedAt: timestamp,
+      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+      extras: extras,
+      _nonce: chatPresenceData.presence?.nonce,
+    };
+  }
+
+  /**
+   * Sets the presence members from the provided presence messages.
+   * @param messages
+   * @private
+   */
+  private _setPresenceMembersFromMessages(messages: PresenceMessage[]): void {
+    this._logger.trace('Presence._setPresenceMembersFromMessages();', { messages });
+    // Clear the current presence members
+    this._presenceMembers.clear();
+    for (const message of messages) {
+      this._setPresenceMember(message);
+    }
+    this._logger.trace('Presence._setPresenceMembersFromMessages(); New presence members set', {
+      members: [...this._presenceMembers.values()],
+    });
+  }
+
+  /**
+   * Processes a presence set change event, updating the presence members list and emitting the appropriate events.
+   * @param presenceSetChange
+   * @private
+   */
+  private _processPresenceSetChange(presenceSetChange: Ably.PresenceSetChange): void {
+    this._logger.trace('Presence._processPresenceSetChange(); Processing new set change event');
+    const { members, current, syncInProgress } = presenceSetChange;
+    // Set the current presence messages
+    this._presenceMessages = members;
+    // Set the sync in progress status
+    this._syncInProgressStatus.previous = this._syncInProgressStatus.current
+    this._syncInProgressStatus.current = syncInProgress;
+
+    // Locate the presence member in the current presence set
+    const presentMember = this._findPresenceMember(current);
+    if (presentMember) {
+      this._handlePresentMember(current, presentMember);
+    } else {
+      this._handleNonPresentMember(current);
+    }
+  }
+
+  /**
+   * Generates a unique key for a presence member based on the clientId and connectionId.
+   * @param clientId
+   * @param connectionId
+   * @private
+   */
+  private _keyFromClientAndConnectionId(clientId: string, connectionId: string): string {
```

Comment:
Make this function take an interface with these properties and just have a single argument? Makes shorthanding the code easier.